hail/src/main/scala/is/hail/expr/ir/agg/StagedBlockLinkedList.scala

```diff
@@ -0,0 +1,236 @@
+    storeDeepCopy(er, typ, Region.loadIRIntermediate(typ)(src), dst)
+
+  // v : Code[T]
+  // dst : Code[PointerTo[T]]
+  private def storeDeepCopy(er: EmitRegion, typ: PType, v: Code[_], dst: Code[Long]): Code[Unit] = {
+    val ftyp = typ.fundamentalType
+    if (ftyp.isPrimitive)
+      Region.storePrimitive(ftyp, dst)(v)
+    else
+      StagedRegionValueBuilder.deepCopy(er, typ, coerce[Long](v), dst)
+  }
+}
+
+class StagedBlockLinkedList(val elemType: PType, val mb: EmitMethodBuilder) {
+  import StagedBlockLinkedList._
+
+
+  /*** State ***/
+
+  val firstNode = mb.newField[Long]
+  val lastNode = mb.newField[Long]
+  val totalCount = mb.newField[Int]
+
+  val storageType = PStruct(
+    "firstNode" -> PInt64Required,
+    "lastNode" -> PInt64Required,
+    "totalCount" -> PInt32Required)
+
+  def load(src: Code[Long]): Code[Unit] = Code(
+    firstNode := Region.loadAddress(storageType.fieldOffset(src, 0)),
+    lastNode := Region.loadAddress(storageType.fieldOffset(src, 1)),
+    totalCount := Region.loadInt(storageType.fieldOffset(src, 2)))
+
+  def store(dst: Code[Long]): Code[Unit] = Code(
+    Region.storeAddress(storageType.fieldOffset(dst, 0), firstNode),
+    Region.storeAddress(storageType.fieldOffset(dst, 1), lastNode),
+    Region.storeInt(storageType.fieldOffset(dst, 2), totalCount))
+
+
+  /*** Temporaries ***
+   * 'i': used to iterate over block buffers in foreachElemAddress
+   * 'p': used to track missingness in foreachBlockNode
+   * 'tmpNode': used to hold pointers to newly allocated nodes */
+  val i = mb.newField[Int]
+  val p = mb.newField[Boolean]
+  val tmpNode = mb.newField[Long]
+
+
+  /*** Node management ***/
+
+  type Node = Code[Long]
+
+  val bufferType = PArray(elemType.setRequired(false), required = true)
+
+  val nodeType = PStruct(
+    "buf" -> bufferType,
+    "count" -> PInt32Required,
+    "next" -> PInt64Optional)
+
+  private def buffer(n: Node): Code[Long] =
+    Region.loadAddress(nodeType.fieldOffset(n, 0))
+
+  private[agg] def capacity(n: Node): Code[Int] =
+    bufferType.loadLength(buffer(n))
+
+  private def count(n: Node): Code[Int] =
+    Region.loadInt(nodeType.fieldOffset(n, 1))
+
+  private def incrCount(n: Node): Code[Unit] =
+    Region.storeInt(nodeType.fieldOffset(n, 1), count(n) + 1)
+
+  private def next(n: Node): Node =
+    Region.loadAddress(nodeType.fieldOffset(n, 2))
+
+  private def hasNext(n: Node): Code[Boolean] =
+    nodeType.isFieldDefined(n, 2)
+
+  private def setNext(n: Node, nNext: Node): Code[Unit] = Code(
+    nodeType.setFieldPresent(n, 2),
+    Region.storeAddress(nodeType.fieldOffset(n, 2), nNext))
+
+  private def initNode(n: Node, buf: Code[Long], count: Code[Int]): Code[Unit] =
+    Code(
+      Region.storeAddress(nodeType.fieldOffset(n, 0), buf),
+      Region.storeInt(nodeType.fieldOffset(n, 1), count),
+      nodeType.setFieldMissing(n, 2))
+
+  private def elemAddress(n: Node, i: Code[Int]): EmitTriplet =
```

Comment:
ah, I get what's happening now---you're not passing the raw value in the EmitTriplet. Would prefer that anything wrapped in an EmitTriplet actually represent a hail value at this point, for minimal confusion---you can do `Region.loadIRIntermediate(eltType)(bufType.elementOffset(...))` here to get the actual value (primitives will return a Code[Int] or whatever for the value). You'll have to adjust the `foreach` functions accordingly.

Reply:
yeah I think that's a good idea